core/src/main/java/com/datastax/oss/driver/internal/core/loadbalancing/helper/InferringLocalDcHelper.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.loadbalancing.helper;
+
+import static com.datastax.oss.driver.internal.core.time.Clock.LOG;
+
+import com.datastax.oss.driver.api.core.config.DefaultDriverOption;
+import com.datastax.oss.driver.api.core.config.DriverExecutionProfile;
+import com.datastax.oss.driver.api.core.metadata.Node;
+import com.datastax.oss.driver.internal.core.context.InternalDriverContext;
+import com.datastax.oss.driver.internal.core.metadata.DefaultNode;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+
+public class InferringLocalDcHelper extends BasicLocalDcHelper {
+
+  public InferringLocalDcHelper(
+      @NonNull InternalDriverContext context,
+      @NonNull DriverExecutionProfile profile,
+      @NonNull String logPrefix) {
+    super(context, profile, logPrefix);
+  }
+
+  /**
+   * This implementation fetches the user-supplied datacenter, if any, from the programmatic
+   * configuration API, or else, from the driver configuration. If no local datacenter is explicitly
+   * defined, this implementation infers the local datacenter from the contact points: if all
+   * contact points share the same datacenter, that datacenter is returned. If the contact points
+   * are from different datacenters, or if no contact points reported any datacenter, an {@link
+   * IllegalStateException} is thrown.
+   *
+   * @return The local datacenter; always present.
+   * @throws IllegalStateException if the local datacenter could not be inferred.
+   */
+  @NonNull
+  @Override
+  public Optional<String> discoverLocalDc() {
+    Optional<String> optionalLocalDc = super.discoverLocalDc();
+    if (optionalLocalDc.isPresent()) {
+      return optionalLocalDc;
+    }
+    Set<String> datacenters = new HashSet<>();
+    Set<DefaultNode> contactPoints = context.getMetadataManager().getContactPoints();
+    for (Node node : contactPoints) {
+      String datacenter = node.getDatacenter();
+      if (datacenter != null) {
```

Comment:
in what case the `node.getDatacenter()` can be `null`?

Reply:
Only in pathological cases, e.g. when `system.peers` contains corrupted data.